### PR TITLE
[ragged-paged-attn] Combine k_pages and v_pages on num_kv_head

### PR DESCRIPTION
**Pull request by bythew3i**
This PR
* Combines k_pages and v_pages on num_kv_head to support sharding num_kv_heads to 1 (multi-chip) while still having performant kernel and scatter.
* Merge `sliding_window` and `soft_cap` change
* Integrate kernel to pytorch cusom kernel
* Refactor tests and improve the tests coverage
* Tested dynamo compilation with dynamic grid from pallas

Tested: 
```
python test/test_pallas.py -v -k PallasTest.test_ragged_paged_attention_wrapper
```

### CODE REVIEW

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py [hunk 1]
```diff
@@ -200,24 +225,28 @@ def ragged_paged_attention_kernel(
...
+    kv_bufs,  # [2, num_kv_pages_per_blk, page_size, num_combined_kv_heads_per_blk, head_dim]
     sems,  # [2, 2]
     l_ref,  # [num_kv_heads_per_blk, num_q_per_blk * num_q_heads_per_kv_head, 128]
     m_ref,  # [num_kv_heads_per_blk, num_q_per_blk * num_q_heads_per_kv_head, 128]
     *,
     sm_scale: float,
-    mask_value: float,
+    sliding_window: int | None = None,
+    soft_cap: float | None = None,
+    mask_value: float | None = DEFAULT_MASK_VALUE,
```
Comment (vanbasten23) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:225]: left some comments in your original [cl](cl/740562542) for the kernel.

Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:225]: Thx! I will resolve there!

File: test/test_pallas.py [hunk 2]
```diff
@@ -762,49 +664,100 @@ def _verify_ragged_paged_attention_with_dynamo(
...
 
     q_xla = q.to("xla")
-    k_pages_xla = k_pages.to("xla")
-    v_pages_xla = v_pages.to("xla")
+    kv_pages_xla = kv_pages.to("xla")
     kv_lens_xla = kv_lens.to("xla")
     page_indices_xla = page_indices.to("xla")
     cu_q_lens_xla = cu_q_lens.to("xla")
     num_seqs_xla = torch.tensor([num_seqs], dtype=torch.int32).to("xla")
+    sliding_window = sliding_window
```
Comment (vanbasten23) [inline test/test_pallas.py:672]: nit no need: line672-673?

File: test/test_pallas.py [hunk 3]
```diff
@@ -762,49 +664,100 @@ def _verify_ragged_paged_attention_with_dynamo(
...
+    soft_cap = soft_cap
+    # Test mask_value
+    mask_value = None
```
Comment (vanbasten23) [inline test/test_pallas.py:675]: imo, we can just use the default mask value rather than letting the user choose one.

**Comment by vanbasten23** [comment]
Mostly LGTM, pending on CI. Thanks Jevin!

File: torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py [hunk 4]
```diff
@@ -593,19 +637,22 @@ def ragged_paged_attention(
   Returns:
     The output of the attention.
   """
-  check_inputs_shapes(q, k_pages, v_pages, kv_lens, page_indices, cu_q_lens,
-                      num_seqs)
-  num_q, num_q_heads, head_dim = q.shape
-  _, page_size, kv_hidden_size = k_pages.shape
-  num_kv_heads = kv_hidden_size // head_dim
+  check_inputs_shapes(q, kv_pages, kv_lens, page_indices, cu_q_lens, num_seqs)
+  if mask_value is None:
```
Comment (vanbasten23) [inline torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:641]: Do you still needed since mask_value is assign to a default value "mask_value: float | None = DEFAULT_MASK_VALUE"

Comment (vanbasten23) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:641]: But you have "= DEFAULT_MASK_VALUE", that means if mask_value is None, it will use DEFAULT_MASK_VALUE, right?

Comment (bythew3i) [reply torch_xla/experimental/pallas_kernels/ragged_paged_attention_v2.py:641]: No, as mentioned, the None in `float | None` is just allowed type, you can do a simple test:

```
>>> def f(a, b: float | None = 1.0):
...     print(a, b)
...
>>> f(2, None)
2 None
```
